internal/sim/corpus.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sim
+
+import (
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// # Simulator Corpus
+//
+// When the simulator runs a failed execution, it persists the failing inputs
+// to disk. The collection of saved failing inputs is called a *corpus*, and
+// each individual entry is called a *corpus entry*. When a simulator is
+// created, the first thing it does is load and re-simulate all corpus entries.
```

Comment:
You mentioned offline that these re-runs may not be very useful to re-run, if the code (and thereby the interleavings) change in the process of fixing the failed runs. I suppose there's some value still, just wanted to get your (updated) opinion.

Reply:
Great question. Here's a recap of my thought process. 

Initially, I wanted to persist the history of a failing execution and replay it step-by-step. If the re-execution diverged from the original execution, I wanted to return an error. Later, I realized that this didn't make any sense because as the user changes their code, the execution will understandably change. 

However, I still think there is benefit to re-executing failing simulations on a best effort basis. For example, if a user encounters a failing simulation, I think it would be very natural for them to add print statements to their code and re-run the simulation. Adding print statements shouldn't cause the simulation to diverge.

I think in the future, we could take this a step further. We could store the history of failing simulations and try to replay them as close to the original execution as possible. As I said above, there may be divergence, but I think there will be cases where we can get close to the original execution. For example, if a user changes how they generate random inputs, this affects the seed of the random number generator which causes the entire simulation to behave differently. However, their code might be the same as before, and we can re-execute the failing simulation perfectly.

Reply:
SGTM!